<div style="text-align: center;">
  <img src="https://github.com/Hack-io-Data/Imagenes/blob/main/01-LogosHackio/logo_naranja@4x.png?raw=true" alt="esquema" />
</div>

# Laboratorio Pandas (EDA y Unión de Datos)

## Los datos

Para este laboratorio, trabajarás con los siguientes conjuntos de datos:

- **`netflix_originals.csv`**: Contiene información sobre producciones originales de Netflix. Las columnas que nos encontraremos en este conjunto de datos son: 

   - `Title`: Contiene el nombre del título de la producción original de Netflix.

   - `Genre`: Especifica el género o combinación de géneros de la producción.

   - `Premiere`: Indica la fecha de estreno de la producción en Netflix.

   - `Runtime`: Contiene la duración de la producción en minutos.

   - `IMDB Score`: Representa la calificación promedio que la producción ha recibido en la plataforma IMDB.

   - `Language`: Especifica el idioma principal o los idiomas en los que se presenta la producción.

- **`netflix_titles.csv`**: Contiene información sobre una variedad de títulos disponibles en la plataforma Netflix, incluyendo tanto películas como series de televisión. Las columnas que nos encontraremos en este conjunto de datos son:

   - `show_id`: Identificador único para cada título en el dataset.

   - `type`: Especifica el tipo de contenido, ya sea "Movie" (película) o "TV Show" (serie de televisión).

   - `title`: Contiene el nombre del título de la producción.

   - `director`: Muestra el nombre del director del título. Puede estar vacío en algunos casos.

   - `cast`: Lista de actores y actrices que participaron en el título. También puede estar vacío en algunos casos.

   - `country`: País o países donde se produjo el título. Esta columna también puede tener valores faltantes.

   - `date_added`: Fecha en la que el título fue añadido al catálogo de Netflix.

   - `release_year`: Año en que el título fue lanzado o producido.

   - `rating`: Clasificación del contenido en términos de audiencia (por ejemplo, PG-13, TV-MA).

   - `duration`: Duración del título. Para películas, se expresa en minutos (e.g., "90 min") y para series en temporadas (e.g., "2 Seasons").

   - `listed_in`: Categorías o géneros bajo los cuales el título está clasificado (por ejemplo, "Documentaries", "TV Dramas").

   - `description`: Breve sinopsis o descripción del contenido del título.

A lo largo de los laboratorios de Pandas, trabajarás con estos conjuntos de datos (o con los derivados del Laboratorio), algunas de las actividades que realizaremos son:

- Exploración de datos y análisis exploratorio (EDA) para familiarizarse con los conjuntos de datos.

- Realización de uniones entre conjuntos de datos, aplicando diferentes técnicas de *merging* y analizando la relevancia de cada método.

- Uso de funciones de selección y filtrado (`loc`, `iloc`) para extraer información clave.

- Desarrollo de análisis descriptivos y visualización de patrones para comprender mejor las decisiones estratégicas de la plataforma.

----

## Objetivo General de la Serie de Laboratorios



1. **Unión de `netflix_titles.csv` con `netflix_originals.csv`:**

   - **Análisis de contenido original:**  Identificarás los títulos que son producciones originales de Netflix y los compararás con los títulos adquiridos. Este análisis te permitirá:

     - Comparar la diversidad de géneros entre los títulos originales y no originales.

     - Evaluar si ciertos géneros o tipos de contenido (como documentales o series) son más frecuentes en las producciones originales.

     - Investigar si la producción original se concentra en determinados países o idiomas.


   - **Evaluación de contenido:** En esta etapa, explorarás si las producciones originales de Netflix tienden a recibir mejores o peores evaluaciones en comparación con el contenido adquirido. Algunas preguntas que se buscarán responder incluyen:

     - ¿Los títulos originales tienen una calificación promedio superior en plataformas como IMDB?

     - ¿Son las producciones originales más consistentes en términos de calidad percibida?

     - ¿Existen diferencias en la popularidad de los títulos originales según su género o país de origen?


   - **Identificación de tendencias:**  Por último, este análisis te permitirá explorar cómo ha evolucionado la estrategia de contenido de Netflix a lo largo del tiempo. Algunos puntos clave a investigar serán:

     - ¿Cómo ha crecido la producción original en comparación con la adquisición de contenido?

     - ¿Qué géneros o tipos de contenido han recibido mayor inversión en los últimos años?
     
     - ¿Existen patrones en los países o idiomas de las producciones originales, especialmente en relación con la expansión global de Netflix?


----


## Ejercicio: Análisis y Enriquecimiento del Catálogo de Netflix

### Parte 1: Análisis Exploratorio de Datos (EDA)

1. **Carga de los datos:**

   - Deberás cargar los dos conjuntos de datos que te hemos proporcionado para realizar este Laboratorio:

     - `netflix_originals.csv`

     - `netflix_titles.csv`

   Deberás revisar la estructura de cada conjunto de datos utilizando los métodos aprendidos durante la lección.

2. **Análisis de columnas y consistencia:**

   - Deberás hacer un análisis exploratorio de cada uno de los conjuntos de datos. Será obligatorio que incluyas las conclusiones generales obtenidas de cada uno de los análisis exploratorios en el Jupyter Notebook de tus soluciones.

   - Explorar qué columnas contiene cada conjunto de datos y cuáles son sus tipos de datos. Identificar si hay valores nulos o duplicados.

   - Algunas preguntas que nos podemos plantear en esta fase son (recuerda que esto es solo un ejemplo, y que podemos explorar todo lo que queramos/necesitemos para entender lo mejor posible los conjuntos de datos):

      - ¿Cuáles son los tipos de datos de cada columna? Nos puede ayudar a detectar si algún tipo de dato está incorrectamente asignado, como fechas almacenadas como cadenas de texto.

      - ¿Existen valores nulos? ¿En qué columnas están? Este paso es crucial para decidir si deben ser eliminados, rellenados o ignorados.

      - ¿Cuántos títulos únicos hay en cada conjunto de datos? Nos puede permitir verificar si existen títulos duplicados que puedan distorsionar el análisis.

      - ¿Qué géneros tenemos en el conjunto de datos? Esto nos puede ayudar a identificar los géneros más populares o incluso a identificar una posible limpieza de dicha columna para poder extraer conclusiones más claras en nuestro análisis.

### Parte 2: Unión de Datos

1. **Realización de la unión:**

   - Deberás realizar la unión que consideres más apropiada entre `netflix_titles.csv` y `netflix_originals.csv`. Además, deberás justificar el tipo de unión que has realizado.

   - Esto nos ayudará a contestar preguntas más adelente del tipo:

     - ¿Cuántos títulos en `netflix_titles.csv` son producciones originales de Netflix?

     - ¿Hay diferencias significativas en la puntuación de IMDB entre títulos originales y no originales?

### Parte 3: Uso de `loc` e `iloc`

Usando el DataFrame creado en la Parte 2: 

1. **Filtrado de datos con `loc`:**

   - Selecciona todas las filas donde el tipo de contenido sea “Movie”.

   - Muestra solo los títulos (title) y la duración (duration) de todos los contenidos que sean de tipo “TV Show”.

   - Selecciona todas las filas donde el país sea “United States”.

   - Muestra todas las películas (type = “Movie”) y selecciona las columnas title y director.

   - Selecciona los títulos (title) y géneros (listed_in) de todas las producciones lanzadas en 2018.

   - Selecciona las filas donde la columna director esté vacía y muestra solo los títulos (title).

2. **Acceso a datos específicos con `iloc`:**

   - Selecciona las primeras 5 filas del conjunto de datos y muestra solo las columnas title, director y country.

   - Muestra las últimas 5 filas del conjunto de datos y selecciona las columnas title y description.

   - Selecciona las primeras 10 filas del dataset y muestra las columnas desde la segunda hasta la quinta (incluidas). 

   - Muestra las últimas 7 filas del dataset y selecciona las columnas title, director, y country usando índices posicionales.

   - Selecciona las filas desde la 15 hasta la 25 (incluidas) y muestra las primeras 4 columnas.

   - Muestra las filas 20 a 30 y solo las columnas title, release_year, y rating usando índices posicionales.


**NOTA IMPORTANTE**: Antes de terminar el lab, guarda el conjunto de datos creado en la fase de unión para trabajar con el en las siguientes lecciones. 

In [2]:
import pandas as pd
import numpy as np

In [3]:
df_netflix = pd.read_csv('datos/netflix_titles.csv',index_col=0)
df_netflixo = pd.read_csv('datos/netflix_originals.csv',index_col=0)

In [ ]:
df_netflixo.head()      # tenemos el Data de el archivo csv netflix original

In [ ]:
df_netflix.head()       # tenemos el Data de el archivo csv netflix titulos

In [8]:
df_netflixo.shape           #Netflix originales 513 filas y 6 columnas


(513, 6)

In [9]:
df_netflix.shape           #Netflix titulos 8807 filas y 12 columnas

(8807, 12)

In [10]:
df_netflixo.duplicated().sum()          #numeros de duplicados en el dataF netfilx original

np.int64(0)

In [11]:
df_netflix.duplicated().sum()                  #numeros de duplicados en el dataF netfilx titulos

np.int64(0)

In [12]:
df_netflixo.dtypes                      #tipos de datos que tenemos en el dataF Netflix original

Title          object
Genre          object
Premiere       object
Runtime         int64
IMDB Score    float64
Language       object
dtype: object

In [13]:
df_netflix.dtypes           #tipos de datos que tenemos en el dataF Netflix titulos

show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

In [14]:
df_netflixo.select_dtypes(np.number).value_counts()  # datos int en DataF netflix orginal y catidad de estos por cada dato

Runtime  IMDB Score
97       6.4           3
92       6.2           3
95       5.8           3
92       5.8           3
94       5.5           3
                      ..
28       6.8           1
         7.0           1
30       6.3           1
         6.9           1
139      6.2           1
Name: count, Length: 471, dtype: int64

In [15]:
df_netflix.select_dtypes(np.number).value_counts()  # datos int en DataF netflix titulos y catidad de estos por cada dato

release_year
2018            1147
2017            1032
2019            1030
2020             953
2016             902
                ... 
1925               1
1947               1
1959               1
1966               1
1961               1
Name: count, Length: 74, dtype: int64

In [16]:
df_netflixo.describe(include= 'O').T        #tipos de datos objet en el dataf netflix originales

count unique                                       top freq
Title      513    513  David Attenborough: A Life on Our Planet    1
Genre      513    106                               Documentary  132
Premiere   513    350                          October 18, 2019    5
Language   513     37                                   English  352

In [17]:
df_netflix.describe(include= 'O').T   #tipos de datos objet en el dataf netflix originales

count unique                                                top  \
show_id      8807   8807                                              s8807   
type         8807      2                                              Movie   
title        8807   8807                                             Zubaan   
director     6173   4528                                      Rajiv Chilaka   
cast         7982   7692                                 David Attenborough   
country      7976    748                                      United States   
date_added   8797   1767                                    January 1, 2020   
rating       8803     17                                              TV-MA   
duration     3994    185                                          2 Seasons   
listed_in    8807    514                       Dramas, International Movies   
description  8807   8775  Paranormal activity at a lush, abandoned prope...   

             freq  
show_id         1  
type         6131  
title           1  
director       19  
cast           19  
country      2818  
date_added    109  
rating       3207  
duration      425  
listed_in     362  
description     4

In [136]:
round(df_netflixo.isnull().sum()/df_netflixo.shape[0]*100,2)    #porcentajes de datos nulos en dataf netflix original

title         0.0
Genre         0.0
Premiere      0.0
Runtime       0.0
IMDB Score    0.0
Language      0.0
dtype: float64

In [137]:

round(df_netflix.isnull().sum()/df_netflix.shape[0]*100,2)       #porcentajes de datos nulos en dataf netflix titulos

show_id          0.00
type             0.00
title            0.00
director        29.91
cast             9.37
country          9.44
date_added       0.11
release_year     0.00
rating           0.05
duration        54.65
listed_in        0.00
description      0.00
dtype: float64

In [57]:
generos_Netflixo = df_netflixo['Genre'].value_counts()
generos_Netflixo                                            #generos y cantidades de estos en netflix original

Genre
Documentary                             132
Drama                                    73
Comedy                                   42
Romantic comedy                          35
Thriller                                 33
                                       ... 
Action-adventure                          1
Concert Film                              1
Animation / Science Fiction               1
Anthology/Dark comedy                     1
Animation/Christmas/Comedy/Adventure      1
Name: count, Length: 106, dtype: int64

In [21]:
df_netflix['type'].value_counts()     #tipos de shows y cantidades en netflix titles

type
Movie      6131
TV Show    2676
Name: count, dtype: int64

In [23]:
df_netflix.info()           #info global de el dataf netflix titles


<class 'pandas.core.frame.DataFrame'>
Index: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      3994 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 894.5+ KB


In [22]:
df_netflixo.info()           #info global de el dataf netflix titles

<class 'pandas.core.frame.DataFrame'>
Index: 513 entries, 0 to 583
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Title       513 non-null    object 
 1   Genre       513 non-null    object 
 2   Premiere    513 non-null    object 
 3   Runtime     513 non-null    int64  
 4   IMDB Score  513 non-null    float64
 5   Language    513 non-null    object 
dtypes: float64(1), int64(1), object(4)
memory usage: 28.1+ KB


In [36]:
df_netflixo.rename(columns={'Title': 'title'}, inplace=True)        #reemplazamos Title por title para unir los dos Datas

In [39]:

df_nuevo = df_netflixo.merge(df_netflix, on="title", how="right")           #unimos los dos datas mediante Titel y que nos deje todos los datos de las 2 tablas 

In [127]:
df_nuevo.head(1)

title        Genre         Premiere  Runtime  IMDB Score  \
0  Dick Johnson Is Dead  Documentary  October 2, 2020     90.0         7.5   

  Language show_id   type         director cast        country  \
0  English      s1  Movie  Kirsten Johnson  NaN  United States   

           date_added  release_year rating duration      listed_in  \
0  September 25, 2021          2020  PG-13   90 min  Documentaries   

                                         description  
0  As her father nears the end of his life, filmm...

In [60]:
movies = df_nuevo.loc[df_nuevo['type'] == 'Movie']      #Seleccionamos los datos que sean peliculas


In [69]:
Tv_shows= df_nuevo.loc[df_nuevo['type']== 'TV Show']        #Seleccionamos dentro de los datos TV las columnas tiempo y los titulos
Tv_shows_duracion= Tv_shows[['Runtime','title']]

In [73]:
df_EEUU = df_nuevo[df_nuevo['country'] == 'United States']  #Sacamos los datos solo del pais EEUU

In [77]:
#Datos solo de peliculas junto al titulo y al director de la pelicula
peliculas_titulo_director = df_nuevo.loc[df_nuevo['type'] == 'Movie', ['title', 'director']]

In [81]:
#datos de producciones en el año 2018 junto con su titulo y tipo de show
producciones18 = df_nuevo.loc[df_nuevo['release_year'] == 2018, ['title', 'listed_in']]


In [87]:
#datos de titulos donde el el dato del director esta vacio
columnasvacias = df_nuevo[df_nuevo['director'].isnull()]['title']

In [93]:
#datos de las 4 primeras filas donde consta titulo,director y pais
titulo_director_con = df_nuevo.loc[:4, ['title', 'director', 'country']]


In [105]:
#datos de las utlimas 5 filas con titulo y descripcion, su resultado no arroja datos
ultimascinco= df_nuevo.loc[:-5, ['title', 'description']]

In [109]:
seleccion_filas_columnas = df_nuevo.iloc[:10, 1:5]

In [117]:
#datos de las ultimas 7 filas con titulo genero y pais
filas7ultimas = df_nuevo.iloc[-7:, [0, 1, 10]]


In [123]:
#datos de las filas 15 a la 25 las 4 primas columnas
datos1= df_nuevo.iloc[14:26, :4]


In [128]:
#Datos obtenidos desde la fila 19 a la 30 incluida con las columnas titulo año de lanzamiento y rating
Datos2 = df_nuevo.iloc[19:31, [0, 12, 13]]


title  release_year rating
19                                           Jaguar          2021  TV-MA
20  Monsters Inside: The 24 Faces of Billy Milligan          2021  TV-14
21                           Resurrection: Ertugrul          2018  TV-14
22                                  Avvai Shanmughi          1996  TV-PG
23     Go! Go! Cory Carson: Chrissy Takes the Wheel          2021   TV-Y
24                                            Jeans          1998  TV-14
25                             Love on the Spectrum          2021  TV-14
26                                   Minsara Kanavu          1997  TV-PG
27                                        Grown Ups          2010  PG-13
28                                       Dark Skies          2013  PG-13
29                                         Paranoia          2013  PG-13
30                                  Ankahi Kahaniya          2021  TV-14

In [129]:
df_nuevo.to_csv('datos/Lab2.csv')  

**Conclusiones del estudio**

En primer lugar comentar que el  estudio partia de unos datos en los que no teniamos datos duplicados, pero si que teniamos datos nulos los cuales serian interesantes de analizar puesto que afectan al director, al pais de origen de la pelicula ,a la duracion de la misma y al ratin. Este aspecto seria conveniente analizar mas adelante con una posible limpieza o sustitucion de los datos.

-La primera observacion que podemos plasmar es que dentro de los 8.806 titulos de la plataforma total de Netfilx solo 512 titulos son originales de esta plataforma, lo que representa 5.8% de los titulos del total.

-Observamos que la plataforma esta en su mayoria completa por titulos con fechas de lanzamientos actuales partiendo estos desde 2016 teniendo su pico en 2018 con 1147 titulos en este año dado. Teniendo estos datos final en el año 2020.

-Los generos mas demandados dentro de la plataforma en la muestra dada son Dramas, International Movies dentro del global de la plataforma y dentro de los titulos originales de esta, lo mas destacado se encuentra en la cateforia documentales.

El idioma que mas destaca es el ingles.

Dentro del total de titulos destaca gran mayoria de peliculas con respecto a TV shows, representando las peliculas un 69,961%.


*Proximos pasos*

Seria interensante un estudio mas detenido de IMDB Score dentro de los generos y los años de las peliculas, para asi saber relacionar la puntuacion con el pais, el genero y la duracion de las peliculas, para poder obtener un resultado en el que podamos saber que tipo de genero y con que duracion de show tiene mas exito y en los paises en los que este exito se tendria.
